File: 프로그래머스/3/214288. 상담원 인원/상담원 인원.cpp
#include <string>
#include <vector>
#include <climits>
#include <queue>
#include <iostream>

using namespace std;

int answer = INT_MAX;

int calcWait(const vector<int>& arr, const vector<vector<int>> reqs) {
    int k = arr.size();
    int totalWait = 0;
    
    vector<vector<pair<int, int>>> byType(k);
    for (const auto& req : reqs) {
        int start = req[0];
        int time = req[1];
        int kind = req[2] - 1;
        
        byType[kind].push_back({start, time});
    }
    
    for (int t = 0; t < k; t++) {
        auto& v = byType[t];
        
        // sort(v.begin(), v.end());
        
        int m = arr[t];
        
        priority_queue<int, vector<int>, greater<>> pq;
        
        for (auto [start, duration] : v) {
            if (pq.size() < m) {
                pq.push(start + duration);
            } else {
                int available = pq.top(); pq.pop();
                
                if (start >= available) {
                    pq.push(start + duration);
                } else {
                    totalWait += available - start;
                    pq.push(available + duration);
                }
            }
        }
    }
    return totalWait;
}

void dfs(int idx, int n, int k, vector<int>& arr, const vector<vector<int>> reqs) {
    if (idx == k - 1) {
        arr[idx] = n;
        for (int x : arr) cout << x << " ";
        cout << "\n";
        answer = min(answer, calcWait(arr, reqs));
        return;
    }
    
    for (int i = 1; i <= n - (k - idx - 1); i++) {
        arr[idx] = i;
        dfs(idx+1, n - i, k, arr, reqs);
    }
}

int solution(int k, int n, vector<vector<int>> reqs) {
    vector<int> arr(k);
    dfs(0, n, k, arr, reqs);
    return answer;
}

File: 프로그래머스/3/214288. 상담원 인원/상담원 인원.java
import java.util.*;

class Solution {
    int k, n;
    int[][] wait;
    int answer = Integer.MAX_VALUE;
    
    public int solution(int k, int n, int[][] reqs) {
        this.k = k;
        this.n = n;
        
        wait = new int[k + 1][n + 1];
        
        List<int[]>[] list = new ArrayList[k + 1];
        
        for(int i = 1; i <= k; i++) {
            list[i] = new ArrayList<>();
        }
        
        for (int[] req : reqs) {
            int start = req[0];
            int time = req[1];
            int type = req[2];
            
            list[type].add(new int[]{start, time});
        }
        
        // wait[type][cnt]
        for (int type = 1; type <= k; type++) {
            for (int cnt = 1; cnt <= n; cnt++) {
                wait[type][cnt] = getWaitTime(list[type], cnt);
            }
        }
        
        dfs(1, n, 0);
        
        return answer;
    }
    
    private void dfs(int type, int remain, int totalWait) {
        if (type == k + 1) {
            if (remain == 0) {
                answer = Math.min(answer, totalWait);
            }
            return;
        }
        
        int leftTypes = k - type; // 3
        
        for (int cnt = 1; cnt <= remain - leftTypes; cnt++) {
            dfs(type + 1, remain - cnt, totalWait + wait[type][cnt]);
        }
    }
    
    private int getWaitTime(List<int[]> reqList, int mentorCount) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for (int i = 0; i < mentorCount; i++) {
            pq.offer(0);
        }
        
        int totalWait = 0;
        
        for (int[] req : reqList) {
            int start = req[0];
            int duration = req[1];
            
            int availableTime = pq.poll();
            
            if (availableTime <= start) {
                pq.offer(start + duration); // 일 줌
            } else {
                totalWait += availableTime - start; // 기다리는 시간
                pq.offer(availableTime + duration);
            }
        }
        
        return totalWait;
    }
}